Implemented Graph Storage V1.0

ksmovlian opened the pull request:
Added required endpoints, working for cases in EXAMPLES.md
Future changes: testing, code comments and performance improvements.

ksmovlian committed:
Implemented Graph Storage V1.0

Added required endpoints, working for cases in EXAMPLES.md
Future changes: testing, code comments and performance improvements.

EvgenKostenko commented:
Read about requirements file and add it please to you repo https://pip.pypa.io/en/stable/user_guide/#requirements-files

EvgenKostenko commented on app.py:
```diff
@@ -0,0 +1,30 @@
+    if requested_nodes_list is None:
+        return jsonify({'error': 'No nodes sent in the request'}), 400
+
+    cyclic_nodes_skipped = graph_storage.add_nodes(requested_nodes_list)
```
Good practice to check `requested_nodes_list` structure 
if I send 
```
{
    "nodes": "hello"
}
```

you app returns 500 error but should 400

EvgenKostenko replied:
you can you https://marshmallow.readthedocs.io for that

EvgenKostenko commented on graph_storage.py:
```diff
@@ -0,0 +1,98 @@
+        node_to_process = self.node_id_to_node.get(new_node_dict['id'])
+        node_to_process_valid_parents = []
+        new_node_parent = new_node_dict.get('parent')
+        if node_to_process is None:
```
better and more clear in python `if node_to_process:` instead of `if node_to_process is None:`
for most you cases it will work fine, but there are some corner cases where it can works incorrectly

EvgenKostenko commented on graph_storage.py:
```diff
@@ -0,0 +1,98 @@
+            })
+        elif node_to_process.parents and new_node_parent not in node_to_process.parents:
+            node_to_process_valid_parents = node_to_process.parents
+            node_to_process.parents += [new_node_parent]
```
better to use `node_to_process.parents.append(new_node_parent)` because in case `+=[]` you create redundant list in memory

EvgenKostenko commented on graph_storage.py:
```diff
@@ -0,0 +1,98 @@
+            node_to_process.parents = node_to_process_valid_parents
+            return False
+
+        if new_node_parent is not None and new_node_parent not in self.node_id_to_node.keys():
```
you should use `new_node_parent not in self.node_id_to_node` without `keys()`
method `keys` return list and search in list take O(n) and search in dict usually O(1)

EvgenKostenko commented on graph_storage.py:
```diff
@@ -0,0 +1,98 @@
+                cyclic_nodes.append(new_node)
+
+        with open('all_nodes.json', 'w') as all_nodes_json:
+            json.dump({'nodes': [node.__dict__ for node in self.node_id_to_node.values()]}, all_nodes_json)
```
user can keep  `self.node_id_to_node.` as dict in your file which allows you serialize you data faster in init stage

EvgenKostenko commented on graph_storage.py:
```diff
@@ -0,0 +1,98 @@
+        are_roots_reached = False
+        all_chains = []
+        node_id_related_chains = []
+        for youngest_node_id in list(self._find_youngest_nodes().keys()):
```
extra casting to `list`

EvgenKostenko commented on graph_storage.py:
```diff
@@ -0,0 +1,98 @@
+            node_to_process.parents += [new_node_parent]
+        else:
+            node_to_process.parents = [new_node_parent]
+            node_to_process.parents = list(filter(None.__ne__, node_to_process.parents))
```
you can just check if its exist before appending it to `node_to_process.parents` list

EvgenKostenko commented on graph_storage.py:
```diff
@@ -0,0 +1,98 @@
+        all_chains = []
+        node_id_related_chains = []
+        for youngest_node_id in list(self._find_youngest_nodes().keys()):
+            all_chains.append([youngest_node_id])
```
Technically we are reading and then building all the chains that we have in our database and basically, it's extra work. We need build only that trees where our node id exists

EvgenKostenko commented on node.py:
```diff
@@ -0,0 +1,23 @@
+    def __repr__(self):
+        return self.node_id
+
+    def is_new_node_cyclic(self, node_id_to_node: dict, node_to_check_id: str):
```
from abstraction perspective this method not related to Node it should be in Graph

EvgenKostenko commented on all_nodes.json:
```diff
@@ -0,0 +1 @@
+{"nodes": [{"node_id": "e12afca2cc0c4d6fb2a27119010f1e20", "parents": ["a0378f4f37c5492091918f53b4da15a5", "c592ff8d473a4924bcd5655ae1c6ffd5"]}, {"node_id": "C90A70BD-23D0-4E8A-BA61-518745390EA5", "parents": ["e12afca2cc0c4d6fb2a27119010f1e20"]}, {"node_id": "D483616F-2A54-497E-96B1-1DE919677000", "parents": ["e12afca2cc0c4d6fb2a27119010f1e20"]}, {"node_id": "final", "parents": ["C90A70BD-23D0-4E ...
```
it's bad practice to commit your test data to the central repository
you can keep int to test

EvgenKostenko commented:
Try to write tests using https://docs.python.org/3/library/unittest.html

ksmovlian committed:
Added first version of requirements.txt;
Implemented validation of the POST request structure using Marshmallow library.
